### PR TITLE
Builtin thread-local object storage for AsyncTask

### PR DESCRIPTION
**Pull request by SOF3**
@dktapps deleted my original pull request. Blame him and ask him for description.

### CODE REVIEW

**Comment by archie426** [comment]
Go for it, nice to see that you are here now! Btw, did u leave the forums?

**Comment by dktapps** [comment]
#blamedktapps
Sorry this has been such a pain in the butt.

**Comment by SOF3** [comment]
Rebased commits

**Comment by Aericio** [comment]
@archie426 He got banned on forums lol

**Comment by SOF3** [comment]
@Aericio Please do not post comments irrelevant to this pull request here.

File: src/pocketmine/scheduler/AsyncTask.php [hunk 1]
```diff
@@ -27,7 +27,7 @@
 /**
  * Class used to run async tasks in other threads.
  *
- * WARNING: Do not call PocketMine-MP API methods, or save objects from/on other Threads!!
+ * WARNING: Do not call PocketMine-MP API methods, or save objects (and arrays cotaining objects) from/on other Threads!!
```
Comment (robske110) [inline src/pocketmine/scheduler/AsyncTask.php]: love this :P

**Review by robske110 — APPROVED** [review]
Awesome :)

**Comment by Aericio** [comment]
> @Aericio Please do not post comments irrelevant to this pull request here. If you continue to post improperly, you may be blocked from @pmmp.

Mind deleting that now? How relevant.

File: src/pocketmine/scheduler/AsyncTask.php (hunk 1 above)
Comment (remotevase) [inline src/pocketmine/scheduler/AsyncTask.php]: Lol hilarious

Comment (SOF3) [inline src/pocketmine/scheduler/AsyncTask.php]: @remotevase please stop spamming on our repo.

Comment (remotevase) [inline src/pocketmine/scheduler/AsyncTask.php]: Okay, but expect to see more pull requests from me that actually contribute CODE not just templates for issues and pull requests.

Comment (SOF3) [inline src/pocketmine/scheduler/AsyncTask.php]: @remotevase Please immediately **stop** using pull requests and issues as chatroom or forums, or respective action will be taken.

**Comment by robske110** [comment]
I could try to test this, if you want
-?

**Comment by robske110** [comment]
The commit name seems to be not very descriptive

**Comment by SOF3** [comment]
@robske110 that's all that this pull request is about. Other changes are just to assist this addition. You don't want to store a value; you just want to retrieve a value, and you do so by storing it first, along with other things like a storage to put these values at. But the main point is that you can use `AsyncTask;:fetchLocal`.

**Comment by robske110** [comment]
well, still looks weird for me, but i guess ok

**Comment by SOF3** [comment]
Tested with this script plugin:

```php
<?php

/**
 * @name AsyncTest
 * @author SOFe
 * @main AsyncTest\AsyncTest
 * @api 2.1.0
 * @version 1.0.0
 */

namespace AsyncTest;

class AsyncTest extends \pocketmine\plugin\PluginBase {
        public function onEnable(){
                $asyncTask = new class($this) extends \pocketmine\scheduler\AsyncTask{
                        public function onRun(){
                                sleep(5);
                        }

                        public function onCompletion(\pocketmine\Server $server){
                                $server->getLogger()->info('$asyncTask');
                                var_dump(get_class($this->fetchLocal($server)));
                        }
                };
                $this->getServer()->getScheduler()->scheduleAsyncTask($asyncTask);

                $asyncTask2 = new class($this) extends \pocketmine\scheduler\AsyncTask{
                        public function onRun(){
                                sleep(10);
                        }

                        public function onCompletion(\pocketmine\Server $server){
                                $server->getLogger()->info('$asyncTask2');
                                // var_dump(get_class($this->fetchLocal($server)));
                        }
                };
                $this->getServer()->getScheduler()->scheduleAsyncTask($asyncTask2);
        }
}
```

Note that anonymous classes are used in this script, and the values to store for an AsyncTask is the `$this` in the `new class($this)` part.

Output with `./start.sh --debug.level=2`:

```php
[12:07:07] [Server thread/INFO]: Enabling AsyncTest v1.0.0
# Irrelevant lines skipped
[12:07:07] [Server thread/INFO]: Done (1.414s)! For help, type "help" or "?"
[12:07:12] [Server thread/INFO]: $asyncTask
~\plugins\async.php:22:
string(19) "AsyncTest\AsyncTest"
[12:07:22] [Server thread/INFO]: $asyncTask2
[12:07:22] [Server thread/DEBUG]: AsyncTask class@anonymous~\plugins\async.php0000006F77B8036A stored local complex data but did not remove them after completion
```

File: src/pocketmine/scheduler/ServerScheduler.php [hunk 2]
```diff
@@ -91,6 +95,59 @@ public function scheduleAsyncTaskToWorker(AsyncTask $task, $worker){
...
+	 *
+	 * @internal Only call from AsyncTask.php
+	 *
+	 * @param AsyncTask $for
+	 *
+	 * @return bool returns false if any data are removed from this call, true otherwise
+	 */	
+	public function removeLocalComplex(AsyncTask $for) : bool{
+		if(isset($this->objectStore[$for])){
+			Server::getInstance()->getLogger()->debug("AsyncTask " . get_class($for) . " stored local complex data but did not remove them after completion");
```
Comment (SOF3) [inline src/pocketmine/scheduler/ServerScheduler.php]: Question: Should this be at the WARN level instead?

**Comment by SOF3** [comment]
The pull request is ready for merge. Apart from the question about WARN vs DEBUG above, it is OK for merging. Please review this pull request if you find no problems in the API design.

**Comment by robske110** [comment]
I think NOTICE would be the right level.

**Review by robske110 — APPROVED** [review]
Looks good now, I'd say ready for merge 👍

File: src/pocketmine/scheduler/AsyncTask.php [hunk 3]
```diff
@@ -42,6 +42,29 @@
 
 	private $crashed = false;
 
+	/**
+	 * Constructs a new instance of AsyncTask. Subclasses don't need to call this constructor unless an argument is to be passed. ONLY construct this class from the main thread.
+	 * <br>
+	 * If an argument is passed into this constructor, it will be stored in a thread-local storage (in ServerScheduler), which MUST be retrieved through {@link #fetchLocal} when {@link #onCompletion} is called.
+	 * Otherwise, a DEBUG level message will be raised and the reference will be removed after onCompletion exits.
```
Comment (dktapps) [inline src/pocketmine/scheduler/AsyncTask.php]: Didn't you change this?

**Review by dktapps — APPROVED** [review]
Discussed in detail.

**Pull request merged**